rbd/watchers_mimic.go

```diff
@@ -59,3 +76,77 @@ func (image *Image) ListWatchers() ([]ImageWatcher, error) {
 	}
 	return imageWatchers, nil
 }
+
+// watchCallbacks tracks the active callbacks for rbd watches
+var watchCallbacks = callbacks.New()
+
+// WatchCallback defines the function signature needed for the UpdateWatch
+// callback.
+type WatchCallback func(interface{})
+
+type watchInstance struct {
+	callback WatchCallback
+	data     interface{}
+}
+
+// Watch represents an ongoing image metadata watch.
+type Watch struct {
+	image   *Image
+	wi      watchInstance
+	handle  C.uint64_t
+	cbIndex int
+}
+
+// UpdateWatch updates the image object to watch metadata changes to the
+// image, returning a Watch object.
+//
+// Implements:
+//  int rbd_update_watch(rbd_image_t image, uint64_t *handle,
+//                       rbd_update_callback_t watch_cb, void *arg);
+func (image *Image) UpdateWatch(cb WatchCallback, data interface{}) (*Watch, error) {
+	if err := image.validate(imageIsOpen); err != nil {
+		return nil, err
+	}
+	wi := watchInstance{
+		callback: cb,
+		data:     data,
+	}
+	w := &Watch{
+		image:   image,
+		wi:      wi,
+		cbIndex: watchCallbacks.Add(wi),
+	}
+
+	ret := C.wrap_rbd_update_watch(
+		image.image,
+		&w.handle,
+		C.callWatchCallback,
+		C.uintptr_t(w.cbIndex))
+	if ret != 0 {
+		return nil, getError(ret)
+	}
+	return w, nil
+}
+
+// Unwatch un-registers the image watch.
+//
+// Implements:
+//  int rbd_update_unwatch(rbd_image_t image, uint64_t handle);
+func (w *Watch) Unwatch() error {
+	if w.image == nil {
+		return ErrImageNotOpen
```

Comment:
I'm not 100% sure I know what you mean, but yes, this code ought to only be triggered by misuse of the api. We've been adding "defenses" like this in other parts of the lib where it is not too inconvenient.

Reply:
Ok, what I meant, that it's even not reachable by misuse, but only by a bug in _this_ library. Is this wrong? If I understand the code correctly, you can only create a `Watch` object with `image != nil`. I'm mainly thinking out loud. What if you want to have tests with 100% code coverage, what would you do with these lines?

Reply:
This is what the first block of the "badWatch" subtest is checking for. It can be triggered by a user of the lib because Watch is a public type:
```golang
w := &cephfs.Watch{}
err := w.Unwatch()
```
The above code will compile. If we left out this check w.Unwatch would panic rather than return an error. To use the watch correctly you need to have an open image and call UpdateWatch to return a well-formed Watch instance. It's a bit "belt and suspenders" but it's a style that the project has evolved into using pretty reguarly.

Reply:
Gotcha, thanks.